### PR TITLE
Add function and move to 0.8.16 solidity

### LINKED ISSUE #1235: Community Pool balance public parameter
`_balanceIsSufficient` is currently a private function.
developers need a public function that can get the balance needed with gasPrice as a parameter.

### PR DESCRIPTION
**Pull request by payvint**

### CODE REVIEW

**Review by sergiy-skalelabs — DISMISSED** [review]

File: proxy/contracts/mainnet/CommunityPool.sol [hunk 1]
```diff
@@ -199,6 +215,26 @@ contract CommunityPool is Twin, ICommunityPool {
...
 
+    /**
+     * @dev Allows `msg.sender` set the amount of gas that should be 
+     * enough for reimbursing any transaction from schain to mainnet.
+     * 
+     * Requirements:
+     * 
+     * - 'msg.sender` must have sufficient amount of ETH on their gas wallet.
+     */
+    function setMultiplier(uint newMultiplierNumerator, uint newMultiplierDivider) external override {
```
Comment (DimaStebaev) [inline proxy/contracts/mainnet/CommunityPool.sol:226]: Add division by zero check.

File: proxy/contracts/mainnet/CommunityPool.sol [hunk 2]
```diff
@@ -212,10 +248,33 @@ contract CommunityPool is Twin, ICommunityPool {
...
+    function getRecommendedRechargeAmount(
+        bytes32 schainHash,
+        address receiver
+    )
+        external
+        view
+        override
+        returns (uint256)
+    {
+        if (_multiplyOnAdaptedBaseFee(minTransactionGas)  <= _userWallets[receiver][schainHash]) {
```
Comment (DimaStebaev) [inline proxy/contracts/mainnet/CommunityPool.sol]: You may calculate it once and remember in variable.

File: proxy/test/CommunityPool.ts [hunk 3]
```diff
@@ -109,6 +109,27 @@ describe("CommunityPool", () => {
...
         });
 
+        it("should recharge wallet if user passed enough money with basefee", async () => {
+            const blockNumber = await ethers.provider.getBlockNumber();
+            const baseFeePerGas = (await ethers.provider.getBlock(blockNumber)).baseFeePerGas;
+            const baseFeeCandidate = (baseFeePerGas ? baseFeePerGas : new BigNumber(1, "hex"));
+            const basefee = baseFeeCandidate.toNumber() === 0 ? BigNumber.from(1) : baseFeeCandidate;
+            const amount = minTransactionGas.mul(basefee);
+            //
+            // comment this check until issue https://github.com/NomicFoundation/hardhat/issues/1688 would be fixed
```
Comment (DimaStebaev) [inline proxy/test/CommunityPool.ts:119]: I'm not sure that this comment has value.

File: proxy/contracts/mainnet/CommunityPool.sol (hunk 1 above)
Comment (payvint) [reply proxy/contracts/mainnet/CommunityPool.sol:226]: Agree will add

File: proxy/contracts/mainnet/CommunityPool.sol (hunk 2 above)
Comment (payvint) [reply proxy/contracts/mainnet/CommunityPool.sol]: Yeap agree

File: proxy/test/CommunityPool.ts (hunk 3 above)
Comment (payvint) [reply proxy/test/CommunityPool.ts:119]: I believe that I can remove this comment and add a check on this function

**Review by DimaStebaev — APPROVED** [review]

**Pull request merged**